### PR TITLE
Fix dataview empty-grid and phantom 'Workflow finished' snackbar

### LINKED ISSUE #3: サブスクリプション機能_事前調査
### Tasks

- [x] サービス比較調査・選定
- [x] 要件整理 -> arayabrain/optinist-for-cloud#2 で対応
  - 要件（必要な機能）
  - システム構成イメージ
  - etc.

### PR DESCRIPTION
**Pull request by milesAraya**
#### Summary
- Fixes issues #1 and #2 from #740, both stemming from the run-result poll and
  the executor observing the same experiment independently.
- **#1 (empty dataview):** a completed run could report FINISHED before the
  `experiment_records` row existed, so the dataview grid was empty until a
  manual reload. The poll now withholds completion until the row is present.
- **#2 (phantom snackbar):** loading/reproducing an already-finished workflow
  fired a "Workflow finished" snackbar even though nothing ran. The snackbar
  now fires only on an in-session run completion.
- This branch is meant to merge into `fix/import-sample-data-menu-740` (which
  carries fixes #3 and #4 of the same issue).

#### Design Decisions
- **#1 — keep the poll read-only, don't make it a second writer.** The issue
  suggested having the poll idempotently register the record itself. A first
  pass did exactly that, but a red-team review found it introduced a real
  data-integrity regression: `experiment_records` has no unique constraint on
  `(workspace_id, uid)`, so the poll and the executor (`snakemake_executor`
  post-process) could both pass an existence check and both `INSERT`,
  producing duplicate rows that later break the `.one()` reads in
  `regist_record_on_workflow_completed` / `update_name`. It also ran thumbnail
  generation on the request thread and let a read-access viewer trigger a
  write.
  - The migration tree currently has three heads, so adding the unique
    constraint that would make an upsert safe is risky and out of scope here.
  - Instead the poll stays read-only: when all nodes are finished but the row
    hasn't landed yet, it reports `completeStatus = processing`, which the
    frontend already treats as "keep polling". The executor remains the sole
    writer; once its row exists the next poll reports completion and the
    dataview opens against a present record. No new writer, no duplicate rows,
    no thumbnail work on the request path, no migration.
  - The check is gated on multiuser mode (where `experiment_records` is used)
    and wrapped in try/except so a read failure can never turn a poll into a
    500 (which the frontend would surface as a phantom "Workflow aborted").
- **#2 — gate the snackbar on the transition, not the value.** Real
  completions always go `START_SUCCESS -> FINISHED` (the poll reducer only
  promotes to FINISHED when the previous status was `START_SUCCESS`). Loading
  or reproducing a finished workflow jumps straight to FINISHED in a single
  reducer, so React never observes `START_SUCCESS` first. Requiring
  `prevStatus === START_SUCCESS` before enqueuing the snackbar suppresses the
  phantom while preserving every genuine completion toast. All other FINISHED
  side-effects (`getExperiments`, lock release) are left unchanged.

### Evidence
- Manual reproduction per the steps below; no automated e2e run recorded in
  this PR (the e2e suite already works around both bugs, see References).

### References
- Issue #740 (items 1 and 2).
- #726 — workflow finalisation race; same root architecture (poll and executor
  observing the same experiment independently). A future consolidation of #726
  could subsume the #1 guarantee.
- e2e suite workarounds this replaces: `waitForDataRows`
  (`frontend/e2e/06-dataview.spec.ts`) for #1, and the phantom-snackbar
  handling in `runTutorial` (`frontend/e2e/helpers.ts`) for #2.

#### Files changed
- `studio/app/common/routers/run.py` -- in `run_result`, hold
  `completeStatus = processing` while the workflow is finished but the
  `experiment_records` row has not been written yet (multiuser only,
  fail-open on error).
- `studio/app/common/core/experiment/experiment_record_services.py` -- add
  `ExperimentRecordService.record_exists()`, a read-only existence check used
  by the poll.
- `frontend/src/store/slice/Pipeline/PipelineHook.ts` -- fire the "Workflow
  finished" snackbar only when the previous status was `START_SUCCESS`.

### Manual Testcases
- **#1:** As a multiuser-mode user, run a workflow to completion. When the run
  spinner clears, open the dataview without reloading. Expected: the finished
  run's record is present (previously the grid could be empty until reload).
- **#1 (negative):** While a workflow is still running, poll responses continue
  to report the run as in progress. Expected: no premature completion.
- **#2:** Reproduce (or load) an experiment that already completed. Expected:
  no "Workflow finished" snackbar appears.
- **#2 (positive):** Start a fresh run and let it finish in-session. Expected:
  the "Workflow finished" snackbar appears exactly once on completion.
- Automated: `pytest studio/tests` and `cd frontend && yarn test` -- add poll
  coverage for the finished-but-record-absent case if the team wants it (see
  Test Coverage below).

### Unit, Integration, Contract Test Coverage
- No new tests added. The backend change reuses already-tested helpers
  (`NodeResult.is_all_nodes_already_finished`, `ExperimentRecordService`), and
  the frontend change is a single guard in an existing effect. A focused
  backend test for `run_result` returning `processing` when the record is
  absent would be the highest-value addition if desired.

#### Difficulties (if any)
- The alembic migration tree has three heads, which ruled out the
  unique-constraint approach for a clean self-healing writer.

#### Risk Assessment
| Area | Risk | Notes |
|------|------|-------|
| `run_result` poll endpoint | Low | Read-only addition, multiuser-gated, wrapped in try/except (fails open to prior behavior). One extra experiment.yaml read per poll near completion. |
| Snackbar effect | Low | Narrows when a toast fires; real completions still go `START_SUCCESS -> FINISHED`. |
| Liveness | Low | If the executor never writes the record (e.g. its post-process throws), the poll would keep reporting `processing` until the user reloads/navigates; this degrades to the original empty-dataview behavior rather than a new data problem. |

### CODE REVIEW

File: studio/app/common/routers/run.py [hunk 1]
```diff
@@ -212,6 +216,28 @@ async def run_result(
...
+        try:
+            if (
+                ExperimentRecordService.is_available()
+                and complete_status != CompleteStatus.PROCESSING
+                and NodeResult.is_all_nodes_already_finished(
+                    ExptConfigReader.read(workspace_id, uid)
+                )
+                and not ExperimentRecordService.record_exists(workspace_id, uid)
+            ):
+                complete_status = CompleteStatus.PROCESSING
```
[the remark this answers is not in the document]
Comment (milesAraya) [reply studio/app/common/routers/run.py]: Leaving as-is. The poll already does ensure_synced_async() + observe() (several file reads) every call, so one extra local yaml parse is marginal, and the expensive record_exists() DB query is already gated last behind is_all_nodes_already_finished. There's no reliable server-side "no pending nodes" signal to gate on, so it'd add complexity for negligible savings.

File: studio/app/common/core/experiment/experiment_record_services.py [hunk 2]
```diff
@@ -21,6 +21,19 @@ def is_available(cls) -> bool:
...
+        with session_scope() as db:
+            return (
+                db.query(ExperimentRecord)
+                .filter(
+                    ExperimentRecord.workspace_id == workspace_id,
+                    ExperimentRecord.uid == unique_id,
+                )
+                .first()
+                is not None
+            )
```
[the remark this answers is not in the document]
Comment (milesAraya) [reply studio/app/common/core/experiment/experiment_record_services.py:35]: Took the lighter query (selecting id only); skipping the int() cast. workspace_id is BIGINT, and MySQL casts the string literal to the column type here, so the index is still used. All six sibling queries in this file also compare it as a str, so casting only here would be inconsistent for no real gain.

File: studio/app/common/routers/run.py [hunk 3]
```diff
@@ -212,6 +216,28 @@ async def run_result(
                 # Create CompleteStatus (converted from RemoteSyncStatus)
                 complete_status = CompleteStatus(sync_status.value)
 
+        # The executor writes the experiment record asynchronously after
+        # completion, so a poll can see all nodes finished before the row lands.
+        # Keep reporting "processing" until it exists, so the dataview isn't
+        # opened against a missing record.
+        try:
```
[the remark this answers is not in the document]
Comment (milesAraya) [reply studio/app/common/routers/run.py]: Done — added test_run_result_record_gate.py with two behavior tests: record absent → completeStatus == "processing", record present → None. Removing the gate makes the absent case fail.

**Review by itutu-tienday — COMMENTED** [review]
## 1. Verdict

**Both fixes are well-designed and, for their intended paths, correct.** The
backend fix (#1) is the highlight: the author correctly *rejected* the naive
"let the poll register the record" approach after a red-team pass found it would
double-`INSERT` into a table with no `UNIQUE(workspace_id, uid)` constraint, and
instead kept the poll strictly read-only and fail-open. That is the right call
and the reasoning in the PR body is sound. The frontend fix (#2) correctly
distinguishes an in-session completion from loading an already-finished workflow
by gating on the `START_SUCCESS → FINISHED` transition.

**I would approve after addressing Concern A (liveness bound) and Concern B
(ERROR masking), or at least explicitly accepting them.** Both are edge cases,
not blockers, but Concern A converts a benign "empty grid until reload" into a
"spinner that never ends," which is a worse failure mode than the one being
fixed. Test coverage is thin (Concern D) — acceptable given the risk profile,
but the frontend guard has zero tests and is the more user-visible change.

---

## 2. Fix #1 — backend poll gate (`run.py`, `experiment_record_services.py`)

### 2.1 Design — correct

```python
try:
    if (
        ExperimentRecordService.is_available()
        and complete_status != CompleteStatus.PROCESSING
        and NodeResult.is_all_nodes_already_finished(
            ExptConfigReader.read(workspace_id, uid)
        )
        and not ExperimentRecordService.record_exists(workspace_id, uid)
    ):
        complete_status = CompleteStatus.PROCESSING
except Exception as e:
    logger.warning(...)
```

Verified against the frontend reducer (`PipelineSlice.ts:57-74`): promotion to
`FINISHED` requires `completeStatus !== PROCESSING`, so forcing `processing`
here reliably holds the poll in `START_SUCCESS` — exactly the "keep polling"
contract the frontend already honours. No new writer, no migration, no thumbnail
work on the request thread. Read-only + multiuser-gated + fail-open is the
right shape, and the `try/except` correctly guarantees a read failure can never
turn a poll into a 500 (which the frontend would surface as a phantom
"Workflow aborted" via `pollRunResult.rejected`).

`record_exists()` is a minimal `SELECT id ... LIMIT 1`-style existence check in
its own `session_scope()` — appropriate and cheap.

### 2.2 Concern A (Medium) — no liveness bound if the executor never writes the record

The gate assumes the executor *will* eventually write the row. It does so for
**both** success and failure runs — `regist_record_on_workflow_completed`
writes `success = workflow_success` even when the workflow failed
(`experiment_record_services.py:74,91`). **But** the executor only calls it when
`observe_success` is true (`snakemake_executor.py:206`), and `observe_success`
is false whenever `WorkflowResult.observe_overall()` throws
(`snakemake_executor.py:197-203`). In that case the row is **never** written.

Consequence: `is_all_nodes_already_finished` is `True` (all nodes are simply
`!= RUNNING` — see `workflow_result.py:429-438`), `record_exists` is
permanently `False`, and the poll reports `processing` **forever**. The run
never reaches `FINISHED`/`ABORTED`, the completion toast never fires, and the
`isRunFinished` lock-release branch in `PipelineHook.ts:240-251` never runs (the
60 s `refreshLock` timer keeps the lock alive as long as the tab is mounted).

The PR's risk table frames this as "degrades to the original empty-dataview
behavior." It is actually **worse**: the old behavior was a *finished* run with
an empty grid that one reload fixed; the new behavior is a perpetual in-progress
spinner. There is no server-side timeout on the poll.

**Recommendation:** bound the hold. E.g. only withhold completion while
`experiment_config.finished_at` is within a short grace window (or for at most N
consecutive polls), then fall through to completion so the run terminates even
if the executor's post-process died. At minimum, `logger.info` when the gate
engages so this state is diagnosable in production.

### 2.3 Concern B (Low–Med) — a terminal `ERROR` upload status gets masked into `processing`

The guard is `complete_status != CompleteStatus.PROCESSING`. If remote upload
failed, `complete_status == CompleteStatus.ERROR`; if the record hasn't landed
yet, the gate rewrites `ERROR → PROCESSING`, hiding a terminal error and
extending polling. Holding makes sense for `None`/`SUCCESS`, but overriding a
terminal `ERROR` is semantically wrong. Suggest:

```python
and complete_status not in (CompleteStatus.PROCESSING, CompleteStatus.ERROR)
```

Low probability (upload-ERROR with record-absent is a narrow window), but the
gate should never resurrect a terminal error into "in progress."

### 2.4 Concern C (Low) — redundant read; risk-table performance claim is inaccurate

- The risk table says "one extra experiment.yaml read per poll **near
  completion**." In fact `ExptConfigReader.read()` runs on **every** poll in
  multiuser mode for the whole duration of a run (`complete_status` is `None`
  while running, so the first three `and` clauses are evaluated every 10 s;
  `is_all_nodes_already_finished` returns `False` until the end, which
  short-circuits `record_exists`, but the YAML read+parse still happens each
  poll). Not just near completion.
- `WorkflowResult.observe()` — called immediately above at `run.py:196` — already
  does `ExptConfigReader.read()` (`workflow_result.py:73`) and already computes
  `is_all_nodes_already_finished` internally (`workflow_result.py:166,178`). The
  new block re-reads and re-computes the identical thing. Consider having
  `observe()` surface "all nodes finished" (and/or the parsed config) so the
  gate reuses it instead of a second read. Minor, but it's duplicated I/O on the
  hot polling path.

---

## 3. Fix #2 — snackbar transition guard (`PipelineHook.ts`)

```ts
} else if (status === RUN_STATUS.FINISHED) {
  if (prevStatus === RUN_STATUS.START_SUCCESS) {
    enqueueSnackbar("Workflow finished", { variant: "success" })
  }
  isRunFinished = true
  dispatch(getExperiments())
}
```

### 3.1 Correctness — verified

- **Genuine in-session run:** `run.fulfilled`/`runByCurrentUid.fulfilled` set
  `START_SUCCESS` (`PipelineSlice.ts:148`); the poll reducer only promotes
  `START_SUCCESS → FINISHED` and there is no intermediate status in between, so
  `prevStatus` is `START_SUCCESS` when `FINISHED` is observed → toast fires. ✓
- **Loading / reproducing a finished workflow:** the `fetchWorkflow.fulfilled`
  matcher sets `START_SUCCESS` and then, in the *same* reducer invocation,
  promotes to `FINISHED` when no nodes are pending (`PipelineSlice.ts:112,
  131-133`). React only ever observes the final `FINISHED`, with `prevStatus`
  being the prior non-`START_SUCCESS` value → no toast. ✓ This is the phantom
  being removed, and the mechanism is exactly as the PR describes.
- Other `FINISHED` side-effects (`getExperiments`, `getAlgoList`, lock release)
  are left intact — correct, only the toast is narrowed.

### 3.2 Note (not a defect)

The fix leans entirely on the invariant "a real completion always passes
through `START_SUCCESS` in a *separate* React commit from `FINISHED`." That
holds today because the poll reducer is the only path that promotes to
`FINISHED` incrementally. If a future change ever lets a fresh run jump to
`FINISHED` in one reducer (as the load path already does), the real toast would
silently disappear. Worth a one-line code comment naming that invariant so it
isn't broken unknowingly. (The existing inline comment explains the *load* case
but not this dependency.)

---

## 4. Test coverage (Concern D)

### Backend — `test_run_result_record_gate.py`
Two behavior tests (record absent → `processing`; record present → not forced).
Reasonable wiring check, but shallow: it mocks both
`is_all_nodes_already_finished` and `record_exists` directly, so it verifies the
`if` is wired, not the predicates. Missing, in decreasing value:

- **Still-running negative:** `is_all_nodes_already_finished=False` must **not**
  force `processing` (guards against the gate firing mid-run).
- **Fail-open path:** `record_exists` raises → `complete_status` preserved and
  no 500 (this is the safety property the whole `try/except` exists for, and
  it's untested).
- **Multiuser-gate off:** `is_available=False` → no behavior change (standalone
  regression guard).
- **Interaction with remote status:** a `SUCCESS`/`ERROR` `complete_status`
  feeding into the gate (directly relevant to Concern B).

### Frontend — none
There is **no** test for the snackbar guard, yet it is the more user-visible
change and the higher regression risk (a wrong guard silently suppresses a real
completion toast). `Pipeline.test.ts` already exercises the reducer; at minimum
add a hook/effect test (or a documented reducer-level assertion) covering:
in-session `START_SUCCESS → FINISHED` fires the toast, and load-path
`* → FINISHED` does not. The PR acknowledges the gap; given the low LOC, one
frontend test is worth adding.

---

## 5. Summary of actions requested

| # | Severity | Item |
|---|----------|------|
| A | **Medium** | Bound the `processing` hold (grace window / max polls) so a dead executor post-process can't strand the run in a permanent spinner; log when the gate engages. (`run.py`) |
| B | Low–Med | Don't override a terminal `ERROR` completeStatus into `processing`: use `complete_status not in (PROCESSING, ERROR)`. (`run.py`) |
| C | Low | Reuse the config/`is_all_nodes_already_finished` already computed by `observe()` instead of re-reading YAML every poll; fix the "near completion" wording in the risk table. |
| D | Low | Add the missing backend cases (still-running, fail-open, gate-off) and at least one frontend test for the snackbar guard. |
| E | Nit | Add a code comment naming the "real completions always transit `START_SUCCESS` in a separate commit" invariant the snackbar guard depends on. |

**Bottom line:** clean, well-scoped PR with a genuinely good backend design
decision (read-only, fail-open, no duplicate-INSERT). The frontend guard is
correct for all current paths. The one thing I'd want changed before merge is a
liveness bound on the `processing` hold (Concern A) — as written, an executor
post-process failure trades an empty-but-finished grid for an unending spinner.

**Review by itutu-tienday — COMMENTED** [review]
## 1. Verdict

The author addressed the substantive concerns well. **A (liveness bound)**,
**B (ERROR masking)**, and **E (invariant comment)** are all correctly fixed,
and the backend test suite grew from 2 → 7 cases covering the paths I asked for.
The two pushbacks (C — reuse `observe()`'s read; D — frontend hook test) are
reasonable and I accept both.

**However, there is one new blocking issue introduced by this change: one of the
seven new backend tests is timezone-dependent and fails in Asia/Tokyo — the
team's own local timezone.** I reproduced it (details in §3). Production code is
correct; the bug is confined to the test's fixture. It must be fixed or CI will
be green while every local `pytest` run in JST is red.

---

## 2. Resolved items — verified in the diff

---

## 3. NEW — Blocking: timezone-dependent test failure

**`test_completes_when_record_absent_past_grace` fails under Asia/Tokyo.**
Reproduced on this machine (JST, `+0900`):

```
TZ=Asia/Tokyo  → 6 passed, 1 FAILED
TZ=UTC         → 7 passed
(ambient JST, no TZ set) → 1 FAILED
```

Failure:
```
test_completes_when_record_absent_past_grace
  assert resp.completeStatus is None
  AssertionError: assert 'processing' is None
```

### Root cause
The test fixture builds `finished_at` in **local** wall-clock:

```python
def _finished_at(seconds_ago):
    return (datetime.now() - timedelta(seconds=seconds_ago)).strftime(DATE_FORMAT)
```

but the mock sets `timezone=None`, so the production helper resolves "now" via
`get_datetime_for_timezone(None)` → `datetime.now(TZ_UTC)` → **UTC**. When the
machine's local zone ≠ UTC, the two clocks differ by the UTC offset. In JST
(+9 h), a `finished_at` 360 s in the local past looks ~9 h in the *future*
relative to UTC-now, so `now − finished` is ≈ −32040 s, which is `< 300` →
`_finished_within_grace` returns `True` → the poll holds `processing` forever in
the test. The assertion expects `None` (past-grace completion) and fails.

The fixture and the code-under-test disagree on which clock "now" is. CI happens
to run in UTC, so the fixture's local clock coincides with the helper's UTC
clock and the bug is invisible there — but it fires on any JST developer's
machine (and would also skew the `finished_seconds_ago=0` "within grace" case,
which only passes by luck because a negative elapsed is still `< 300`).

### Fix (pick one)
- **Simplest:** build the fixture timestamp in UTC to match the `timezone=None`
  path — `datetime.now(timezone.utc).replace(tzinfo=None)` in `_finished_at`; or
- Give the mock an explicit `timezone` and build `_finished_at` in that same zone
  (`get_datetime_for_timezone(tz)`), so fixture and helper share one clock; or
- Freeze time (`freezegun` / patch `run.get_datetime_for_timezone`) so the test
  is independent of the host clock entirely — most robust, since the other
  time-based assertions (`within grace`, `seconds_ago=0`) are currently only
  passing by the sign of the skew.

This is purely a test-hermeticity bug; the production grace logic is sound.

---

## 4. Accepted pushbacks

- **C (reuse `observe()`'s config read):** Accepted. The `record_exists` DB query
  is still gated last behind `is_all_nodes_already_finished`, and one extra local
  YAML parse on a path that already does `ensure_synced_async` + `observe()` is
  marginal. Restructuring `observe()`'s return contract for that saving isn't
  worth it. Fine to decline.
- **D (frontend hook test):** Accepted. The reducer transition is already covered
  in `Pipeline.test.ts`, and the effect guard is one line now documented by the
  (E) invariant comment. Mocking snackbar/dispatch/lock for a `useRunPipeline`
  effect test is low value for the setup cost.

---

## 5. Summary

| # | Prior severity | Status |
|---|----------------|--------|
| A | Medium | ✅ Fixed — 300 s grace bound, warns and completes past window |
| B | Low–Med | ✅ Fixed — `not in (PROCESSING, ERROR)` |
| E | Nit | ✅ Fixed — invariant comment added |
| D (backend) | Low | ✅ Fixed — 7 cases |
| C | Low | ⤴️ Declined — reasonable |
| D (frontend) | Low | ⤴️ Declined — reasonable |
| **NEW** | **Blocking** | ❌ `test_completes_when_record_absent_past_grace` fails under JST — non-hermetic time fixture (see §3) |

**Bottom line:** the production change is now in good shape — the liveness bound
was the important one and it's done correctly, with production timezone handling
that is genuinely offset-safe. The only thing standing in the way is a
timezone-dependent test that is red on JST machines (green on UTC CI). Fix the
fixture's clock (§3) and this is an approve.

**Review by itutu-tienday — APPROVED** [review]
